[SPARK-19382][ML]:Test sparse vectors in LinearSVCSuite

wangmiao1981 opened the pull request:
## What changes were proposed in this pull request?

Add unit tests for testing SparseVector.

We can't add mixed DenseVector and SparseVector test case, as discussed in JIRA 19382.

 def merge(other: MultivariateOnlineSummarizer): this.type = {
if (this.totalWeightSum != 0.0 && other.totalWeightSum != 0.0) {
require(n == other.n, s"Dimensions mismatch when merging with another summarizer. " +
s"Expecting $n but got $
{other.n}

.")


## How was this patch tested?

Unit tests

wangmiao1981 committed:
unit test backup

wangmiao1981 committed:
add SparseVector test

SparkQA commented:
**[Test build #72302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72302/testReport)** for PR 16784 at commit [`fc1f7d1`](https://github.com/apache/spark/commit/fc1f7d10134638dfe5130eb19784852207acebd5).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72302/
Test FAILed.

wangmiao1981 commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #72312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72312/testReport)** for PR 16784 at commit [`fc1f7d1`](https://github.com/apache/spark/commit/fc1f7d10134638dfe5130eb19784852207acebd5).

SparkQA commented:
**[Test build #72312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72312/testReport)** for PR 16784 at commit [`fc1f7d1`](https://github.com/apache/spark/commit/fc1f7d10134638dfe5130eb19784852207acebd5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72312/
Test FAILed.

wangmiao1981 commented:
retest this please

wangmiao1981 commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #72314 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72314/testReport)** for PR 16784 at commit [`fc1f7d1`](https://github.com/apache/spark/commit/fc1f7d10134638dfe5130eb19784852207acebd5).

SparkQA commented:
**[Test build #72314 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72314/testReport)** for PR 16784 at commit [`fc1f7d1`](https://github.com/apache/spark/commit/fc1f7d10134638dfe5130eb19784852207acebd5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72314/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -220,12 +246,13 @@ object LinearSVCSuite {
 
-    // Generate noisy input of the form Y = signum(x.dot(weights) + intercept + noise)
+  // Generate noisy input of the form Y = signum(x.dot(weights) + intercept + noise)
   def generateSVMInput(
```
This API is strange, where the caller expects numFeatures = weights.size, but really numFeatures = 10 * weights.size if isDense=false.  Please update it to construct a random dense or sparse vector first (both of length weights.size) and then compute y to make the API more consistent.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -234,7 +261,12 @@ object LinearSVCSuite {
       if (yD > 0) 1.0 else 0.0
     }
-    y.zip(x).map(p => LabeledPoint(p._1, Vectors.dense(p._2)))
+    val index = (0 to weights.length - 1).toArray
```
Move inside if-then to branch where it is used.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -203,6 +227,8 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     val svm = new LinearSVC()
     testEstimatorAndModelReadWrite(svm, smallBinaryDataset, LinearSVCSuite.allParamSettings,
       checkModelData)
+    testEstimatorAndModelReadWrite(svm, smallSparseBinaryDataset, LinearSVCSuite.allParamSettings,
```
No need for this.  Once the model has been fit, its training data is irrelevant.

jkbradley reviewed (commented):
@wangmiao1981  Thanks for the patch!  I don't think that we need to replicate all tests using dense and sparse vectors.  The main concern is making sure that sparse and dense vectors produce the same behavior.

Is this multivariate online summarizer issue really a bug?  Or is it from passing in sparse vectors which are 10x longer than the dense vectors?

I'd recommend:
* Change the dense and sparse datasets so that they match:
  * create the sparse datasets
  * convert the sparse ones to dense using Vector.toDense
* Create a test suite which trains on the 2 datasets and confirms that the models produced are exactly the same.
* Revert the other test changes.  Unless we can think of reasons why the aspect being tested will interact with vector sparsity, then we don't need to add tests.

wangmiao1981 committed:
address review comments

SparkQA commented:
**[Test build #73605 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73605/testReport)** for PR 16784 at commit [`2422f08`](https://github.com/apache/spark/commit/2422f08a933ad3c57f60bb5c1148d7a3ab3c8ba2).

wangmiao1981 committed:
revert one test

SparkQA commented:
**[Test build #73606 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73606/testReport)** for PR 16784 at commit [`8ff2709`](https://github.com/apache/spark/commit/8ff2709a1ffa9856fd598d7447339ffe33103176).

wangmiao1981 commented:
@jkbradley I simplified the tests and modified the data generation API by using toSparse method, which eliminates the index variable.
"Is this multivariate online summarizer issue really a bug? Or is it from passing in sparse vectors which are 10x longer than the dense vectors?"
It is not a bug based on my understanding. It is because the sparse size is larger than the dense vector size, which means we can't mix using them together.

wangmiao1981 commented:
@jkbradley I simplified the test cases and modified the data generation API using to toSparse method, which eliminates the index variable. 
"Is this multivariate online summarizer issue really a bug? Or is it from passing in sparse vectors which are 10x longer than the dense vectors?"
I think it is not a bug, which is due to the different sizes between dense and sparse vectors. It means that we can't mix using the two together.

SparkQA commented:
**[Test build #73605 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73605/testReport)** for PR 16784 at commit [`2422f08`](https://github.com/apache/spark/commit/2422f08a933ad3c57f60bb5c1148d7a3ab3c8ba2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73605/
Test PASSed.

SparkQA commented:
**[Test build #73606 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73606/testReport)** for PR 16784 at commit [`8ff2709`](https://github.com/apache/spark/commit/8ff2709a1ffa9856fd598d7447339ffe33103176).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73606/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -68,13 +77,21 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     assert(model.transform(smallValidationDataset)
       .where("prediction=label").count() > nPoints * 0.8)
+    val sparseModel = svm.fit(smallSparseBinaryDataset)
+    assert(sparseModel.transform(smallSparseValidationDataset)
```
No need to do transform for this model; calling checkModels should suffice.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -68,13 +77,21 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     assert(model.transform(smallValidationDataset)
       .where("prediction=label").count() > nPoints * 0.8)
+    val sparseModel = svm.fit(smallSparseBinaryDataset)
+    assert(sparseModel.transform(smallSparseValidationDataset)
```
same here

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala:
```diff
@@ -51,6 +54,12 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     binaryDataset = generateSVMInput(1.0, Array[Double](1.0, 2.0, 3.0, 4.0), 10000, 42).toDF()
+
+    // Dataset for testing SparseVector
+    smallSparseBinaryDataset = generateSVMInput(A, Array[Double](B, C), nPoints, 42, false).toDF()
```
Why call generateSVMInput again?  It seems brittle.  I'd prefer to use a UDF to convert the vectors to sparse vectors here.

jkbradley reviewed (commented):
Thanks for the simplification!  A few comments.

wangmiao1981 committed:
simplify data generation

wangmiao1981 committed:
remove blank lines

SparkQA commented:
**[Test build #73630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73630/testReport)** for PR 16784 at commit [`911f3ba`](https://github.com/apache/spark/commit/911f3ba8a023ec79392a4e6cec9e5bea23ed6838).

SparkQA commented:
**[Test build #73632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73632/testReport)** for PR 16784 at commit [`2a012fa`](https://github.com/apache/spark/commit/2a012fa64d5f4fc3baf8d58bee27836a3e6a1268).

SparkQA commented:
**[Test build #73630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73630/testReport)** for PR 16784 at commit [`911f3ba`](https://github.com/apache/spark/commit/911f3ba8a023ec79392a4e6cec9e5bea23ed6838).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73630/
Test PASSed.

SparkQA commented:
**[Test build #73632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73632/testReport)** for PR 16784 at commit [`2a012fa`](https://github.com/apache/spark/commit/2a012fa64d5f4fc3baf8d58bee27836a3e6a1268).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73632/
Test PASSed.

jkbradley commented:
Thanks for the updates.  This LGTM pending the conflict resolution.  Sorry for the delay!

wangmiao1981 committed:
Merge branch 'master' into bk

wangmiao1981 commented:
Resolved. Thanks!

SparkQA commented:
**[Test build #74033 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74033/testReport)** for PR 16784 at commit [`1f98622`](https://github.com/apache/spark/commit/1f98622e5cda7e0c47b4c6eab75463dece272e15).

SparkQA commented:
**[Test build #74033 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74033/testReport)** for PR 16784 at commit [`1f98622`](https://github.com/apache/spark/commit/1f98622e5cda7e0c47b4c6eab75463dece272e15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FPGrowth @Since(\"2.2.0\") (`
  * `  class FPGrowthModelWriter(instance: FPGrowthModel) extends MLWriter `
  * `case class ResolveInlineTables(conf: CatalystConf) extends Rule[LogicalPlan] `
  * `abstract class CSVDataSource extends Serializable `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74033/
Test PASSed.

jkbradley commented:
LGTM
Thanks @wangmiao1981 !
Merging with master
